### DISCUSSION TITLE
Feature: add/subtract every row of matrix by vector

### DISCUSSION THREAD

**Discussion by JasonShin** [josdejong/mathjs#2753] [Ideas]
I am getting a dimension error while using math.subtract()

```
const X = [[1, 2], [3, 4], [5, 6]]
const A = math.transpose(X);
// A = [ [ 1, 3, 5 ], [ 2, 4, 6 ] ] 
const M = math.mean(A, 1);
// M = [ 3, 4 ]
const C = math.subtract(X, M);
```

Running the code raises the below error

```
....mathjs/lib/function/arithmetic/subtract.js:172
    throw new DimensionError(xsize.length, ysize.length);

```

I believe `math.subtract(X, M)` should return

```
[[-2., -2.],
[ 0.,  0.],
[ 2.,  2.]]
```

same goes for `math.sum` `math.multiply`

**Comment by JasonShin** [comment]
In the meantime, I will write an extension method to achieve my goal

Temporary extension method that I am using:

```
/**
 * Subtracts two matrices
 * @param X
 * @param y
 */
const subtract = (X, y) => {
  const _X = _.clone(X);
  for (let rowIndex = 0; rowIndex < _X.length; rowIndex++) {
    const row = X[rowIndex];
    for (let colIndex = 0; colIndex < row.length; colIndex++) {
      const column = row[colIndex];
      // Supports y.length === 1 or y.length === row.length
      if (y.length === 1) {
        const subs = y[0];
        _X[rowIndex][colIndex] = column - subs;
      } else if (y.length === row.length) {
        const subs = y[colIndex];
        _X[rowIndex][colIndex] = column - subs;
      } else {
        throw Error(`Dimension of y ${y.length} and row ${row.length} are not compatible`);
      }
    }
  }
  return _X;
}
```

**Comment by josdejong** [comment]
Yes, that can be a useful extension. Right now `math.subtract`, `math.add`, etc simply do element-wise subtraction/addition. Allowing to subtract a vector row-wise from a matrix is something you can do in Matlab and Octave. 

I will change the title of this topic a bit.

Anyone interested in implementing this feature?

**Reply by dvd101x** [reply]
Hi Jos. I’m interested in this topic and I’m familiar with numpy’s and matlab’s behavior 

I think I can make an algorithm that might aid in the implementation of this feature.

**Reply by dvd101x** [reply]
I have something ready for broadcasting matrices of any dimensions, I'm working on some edge cases, maybe I can share something in a week or two.

**Reply by josdejong** [reply]
Thanks @dvd101x ! 

So, will your approach be enabling a syntax like `subtract(A, dimension)`, similar to statistics functions like `mean`? I think that will be a good approach.

**Reply by dvd101x** [reply]
Hi Jos, currently I have something that broadcasts two or more matrices like `broadcast(A, B)` in N-dimensions. Once the inputs become the same size; all functions like `math.add(A, B)`, `math.subtract(A, B)`, `math.multiply(A, B)` and `math.divide(A, B)` could work as they currently support matrices of the same size.

My intention is modifying add, subtract, multiply and divide thus that they broadcast if needed and then work as they do now.

I think I can make a proof of concept that works in the browser by tweaking these functions, sadly I'm not familiar with building from source and other good practices that would be needed.

**Reply by dvd101x** [reply]
Additional to that I will be testing the following

- To include include `math.typed( )` for inputs as I was doing it more complex
- To make it such that two or more arguments like `broadcast(A, B, C)` can be used in `add(A, B, C)` and `dotMultiply(A, B, C)` if needed
- To make it such that exactly two arguments like `broadcast(A, B)` can be used in `subtract(A, B)` and `dotDivide(A, B)` if needed

**Reply by josdejong** [reply]
Is your `broadcast` solution similar as the [`reduce` function](https://github.com/josdejong/mathjs/blob/develop/src/utils/collection.js#L67-L88) used by for example `mean(A, dimension)` but then applicable to two matrices instead of one? What API do you have in mind in the end for the functions add, multiple, etc, to apply the operation on a specific dimension?

**Reply by dvd101x** [reply]
Hi,

It's similar in principle to the `reduce` but it works in the oposite way by using `concat`. I need to rename some stuff to be more readable as the reference provided.

I'm testing two functions

## 1.- `broadcast_to(A, dimensions)`

Converts `[3, 4]` to dimensions `[3, 2]` an returns the matrix broadcasted to the required dimensions `[[3,4], [3,4], [3,4]]`.

## 2.- `broadcast_matrices(A, B)`

Broadcast two or more matrices and returns all the inputs broadcasted

For example if `A = [[1,2], [3,4], [5,6]]` and `B = [3,4]`
- Calculates the dimension to broadcast `[3, 2]`
- Checks that `[[1,2], [3,4], [5,6]]` has already `[3, 2]` dimensions and is left as such
- Broadcasts `[3,4]` to dimensions `[3, 2]` as such `[[3,4], [3,4], [3,4]]`

After that process of broadcasting most operations can work like `math.add(A, B)`, `math.subtract(A, B)`, `math.dotMultiply(A, B)` and `math.dotDivide(A, B)`

Here is an example similar to @JasonShin request where B has to be internally broadcasted.

$$
\begin{align} 
A &= \begin{bmatrix}1 & 2 \\\ 3 & 4 \\\ 5 & 6 \end{bmatrix} \\
B &= \begin{bmatrix} 3 & 4 \end{bmatrix} \\ 
C = A - B &= \begin{bmatrix}1&2\\\3&4\\\5&6\end{bmatrix} - \begin{bmatrix} 3 & 4 \\\ 3 & 4 \\\ 3 & 4 \end{bmatrix} \\
C &= \begin{bmatrix} -2 & -2 \\\ 0 & 0 \\\ 2 & 2 \end{bmatrix}
\end{align}
$$

I think there might be no need to include additional functions into the API, I'm testing an example of importing new add, subtract, multipy and divide functions that broadcast if needed.

In my mind many operations already broadcast in mathjs, except it only broadcasts between a scalar and a matrix/vector, like `2+[1, 2]`. Maybe if this works, it could help to broadcast between N-dimensions that are possible to broadcast.

**Reply by dvd101x** [reply]
An explanation of how this could work inside `math.evaluate( )`

```python
A = [1, 2; 3, 4; 5, 6];
M = mean(A, 1);
M = [M, M, M];      # broadcast would do something like this
C = A - M           # yields [[-2, -2], [0, 0], [2, 2]]
```
And once broadcasting is implemented the original expression could be reduced like this.

```python
A = [1, 2; 3, 4; 5, 6];
C = A - mean(A, 1)   # could yield [[-2, -2], [0, 0], [2, 2]]
```

**Reply by dvd101x** [reply]
I have an example of the algorithm working like so:

```python
A = [1, 2; 3, 4; 5, 6];
C = A - mean(A, 1)   # yields [[-2, -2], [0, 0], [2, 2]]
```
To validate the behavior, please copy paste the the above example into the left pane of this project
https://dvd101x.github.io/web-thermodynamics/

I included broadcasting between scalars, vectors and N dimensional matrixes to `math.add( )` , `math.subtract( )`, `math.dotMultiply( )` and `math.dotDivide( )`.

I will be looking into ways of including `math.typed( )` instead of some functions I made to check the types and try to make it easier to read.

> **Note:** I edited some errors in my comments regarding `divide`, when I should have used `dotDivide`

> **Off topic:** I will be looking into other functions that could benefit from elementwise operations with two or more inputs like log(N, base) that might be nice to accept broadcasting.

**Reply by josdejong** [reply]
### 1. broadcast_to(A, dimensions)

So it is a sort of a repeat function. I guess that can be useful to create a filled matrix with a certain repeated pattern.

### 2. broadcast_matrices(A, B)
 
Ahh, now I get how you can use broadcasting to make this work. It is a very neat solution. Technically, I think it is not the fastest, temporarily creating a large broadcasted matrix. But we can start with this, and if needed, we can always switch to a faster solution similar to the `reduce` that we have for `mean` and other functions. 

This functionality will be very useful, I'm looking forward to have this working! 

We can go through [the list with functions](https://mathjs.org/docs/reference/functions.html) so see which functions can benefit from this: search for all text occurrences of `(x, y)` and `(a, b)` on the page :grin: 

### 3. add(a, b, dimension)

One question: suppose we have a square matrix, and we're going to broadcast a vector. Which dimension should we take, and should we make this customizable, like with `mean` you can provide the dimension as second argument? I.e. should we implement `add(a, b, dimension)`?

**Reply by dvd101x** [reply]
Thank you for reviewing this, I hope you don't mind but here is an overview of the algorithm.

# broadcast sizes

Given a set of matrix sizes, find the broadcasted size that can fit all dimensions for all matrices. All matrices would need to check if they need to grow, for that it's need to increase size empty or 1 to the maximum of all sizes in that dimension.

matrix | $dim_1$ | $dim_2$ (rows) | $dim_3$ (columns)|
-------|----|-----|----|
A      | 2  |3    |4|
B      |    | 3   | 1|
C      |    |1    |4
broadcasted size | 2 | 3 | 4|

Thus is only possible to broadcast a dimension that is empty or 1 to a higher size.

matrix | $dim_1$ | $dim_2$ (rows) | $dim_3$ (columns)|
-------|----|-----|----|
A | 2 |3|4|
B |  | 3 | 1|
C | 1 | ~~2~~ |4
broadcasted size | 2 | **fail** | 4|

> **error:** Failed because it's not possible to broadcast shape 2 to shape 3. For this to work that dimension should be empty or 1

The algorithm must work for **N** dimensions and **M** matrices

## 1. broadcast_to(A, dimensions)

Yes, it first checks if it's "broadcastable" and then repeats the numbers, the issue here is that it doesn't repeat a patter as it's not supposed to broadcast an existing dimension that is not of size one.

Personally I think it might be useful for something like `[1, 2] + [3, 4, 5, 6]` as `[1, 2, 1, 2] + [3, 4, 5, 6]` and which could yield `[4, 6, 6, 8]` but the typical convention is that it doesn't work like that and should throw an error of not matching dimensions so I kept it that way.

For something like patterns we can use orthogonal dimensions and then reshaping back

```python
A = [1, 2] + [[3,4],[5,6]]   # [[4, 6], [6, 8]]
reshape(A, [-1])             # [4, 6, 6, 8]
```

## 2. broadcast_matrices(A, B)

I agree it's not very efficient in that sense, it was easier for me to lean on the capabilities to do operations on matrices of the same size. Maybe a better way would be to modify the code that does operations on matrices of the same size to first check for sizes, calculate only the broadcasted size for all matrices and if it doesn't throw an error then use the modulus operator to find the correct indexes of the matrices, like `math.mod(indexPositionInThisDimension, broadcastedSizeInThisDimension)` in the case of *0-indexing* in a way to optimize for memory, maybe affecting a bit of performance. Maybe just checking if the broadcasted index in the dimension is greater than 1 on the matrix then use it, otherwise use index 0 depending on what is more computationally expensive, a *mod* or an *if*. I haven't thought about this thoroughly. 

I can check on the list of functions for candidates.

## 3. add(a, b, dimension)

I think there is no need to specify a dimension as it takes the already available dimensions for all matrices and the matrices can be reshaped beforehand.

In the example of a square matrix and a vector

The dimension that gets repeated is the first one (counting on the final broadcasted size) and one that does not exist yet on the vector.

matrix | $dim_1$ | $dim_2$|
------|-----|----|
A |2|2|
B | | 2|
broadcasted size | 2 | 2|

If we want to repeat on the second one we could first `reshape(B, [-1, 1])`

matrix | $dim_1$ | $dim_2$|
------|-----|----|
A |2|2|
B |2|1|
broadcasted size | 2 | 2|

**Reply by dvd101x** [reply]
Hi, I found these candidates for broadcasting:

- add
- dotDivide
- dotMultiply
- dotPow
- gcd
- lcm
- log: maybe
- floor
- round
- subtract
- bitAnd
- bitOr
- bitXor
- and
- or
- xor
- compare
- compareText
- deepEqual
- equal
- equalText
- larger
- largerEq
- smaller
- smallerEq
- unequal
- atan2: it has not lost it's elemen twise operations
- to

>  Method: for every function with multiple inputs, check if the second argument could be a matrix and operations are element wise

**Comment by JasonShin** [comment]
Once I finish the `getMatrixDataType` pull request, I am happy to implement it =). Sorry I was busy for the last few days =)...

**Comment by josdejong** [comment]
:+1: thanks for the offer. and for keeping focus :)

**Comment by josdejong** [comment]
We can consider two cases in this regard, as noted by @alvinhui in #1228:

> ## Case 1:
> 
> ```javascript
> const a = [[1, 2, 3], [4, 5, 6], [7, 8, 9]];
> const b= [[9], [8], [7]]; 
> 
> math.subtract(a, b); // should be [[-8, -7, -6],  [-4, -3, -2], [ 0,  1,  2]]
> ```
> 
> ## Case 2
> 
> ```javascript
> const a = [[1, 2, 3], [4, 5, 6], [7, 8, 9]];
> const b= [9, 8, 7];
> 
> math.subtract(a, b); // should be [[-8, -6, -4], [-5, -3, -1], [-2,  0,  2]]
> ```
> 
> This behavior is similar to subtraction between Python arrays:
> 
> ```python
> a = np.array([[1,2,3], [4,5,6], [7,8,9]])
> b = np.array([9,8,7])
> c = [[9], [8], [7]]; 
> 
> a - b
> a - c
> ```

**Reply by dvd101x** [reply]
Hi Jos,

I have an algorithm that could work for both cases. To validate the algorithm please copy paste the the below example into the left pane of this project
https://dvd101x.github.io/web-thermodynamics/

```python
# ## Case 1
# 
# $$ \begin{align}
# a &= \begin{bmatrix}
# 1 & 2 & 3 \\
# 4 & 5 & 6 \\
# 7 & 8 & 9
# \end{bmatrix},
# b = \begin{bmatrix}9\\8\\7\end{bmatrix} \\
# a - b &= \begin{bmatrix}
# 1 & 2 & 3 \\
# 4 & 5 & 6 \\
# 7 & 8 & 9
# \end{bmatrix}
# - \begin{bmatrix}
# 9 & 9 & 9 \\
# 8 & 8 & 8 \\
# 7 & 7 & 7
# \end{bmatrix}
# = \begin{bmatrix}
# -8 & -7 & -6 \\
# -4 & -3 & -2 \\
# 0 & 1 & 2
# \end{bmatrix}
# \end{align}$$

a = [[1, 2, 3], [4, 5, 6], [7, 8, 9]];
b = [[9], [8], [7]]; 
a - b

# ## Case 2
# 
# $$ \begin{align}
# a &= \begin{bmatrix}
# 1 & 2 & 3 \\
# 4 & 5 & 6 \\
# 7 & 8 & 9
# \end{bmatrix},
# b = \begin{bmatrix}9 & 8 & 7\end{bmatrix} \\
# a - b &= \begin{bmatrix}
# 1 & 2 & 3 \\
# 4 & 5 & 6 \\
# 7 & 8 & 9
# \end{bmatrix}
# - \begin{bmatrix}
# 9 & 8 & 7 \\
# 9 & 8 & 7 \\
# 9 & 8 & 7
# \end{bmatrix}
# = \begin{bmatrix}
# -8 & -6 & -4 \\
# -5 & -3 & -1 \\
# -2 & 0 & 2
# \end{bmatrix}
# \end{align}$$

a = [[1, 2, 3], [4, 5, 6], [7, 8, 9]];
b = [9, 8, 7];
a - b
```

> **notes:**
> - the comments do nothing and just render in markdown+latex in that project.
> - it's supposed to work in n-dimensions, thus many more examples can be tested

**Reply by josdejong** [reply]
### 1. broadcast_to(A, dimensions)

Yes agree, we should be a bit conservative in "magically" broadcasting operations to make it work when the dimensions do not match. That will be quite confusing for the user and could cause users to not spot an error.

So, I think we should not apply broadcasting for `[1, 2] + [[3,4],[5,6]]`, but we can support it explicitely like `repeat([1, 2], 2) + [[3,4],[5,6]]` or something like that (the name can be `repeat` or `broadcast` or something like that).

### 2. broadcast_matrices(A, B)

:thumbsup: 

We can first just go with the simplest approach now, and only look into a more advanced solution to make it faster in a second iteration. Let's start simple.

Thanks for working out the list with candidates. Do you think it would be best do directly implement support for this for all these functions? To keep things manageable, it may be good to implement support for just a few functions (like `add` and `subtract`) in the first PR, and expand on that in a second PR? 

### 3. add(a, b, ~dimension~)

You're right. I was confused by how this functionality works for `mean`, but that is different. For `add` and `subtract` with differently sized matrices, you can indeed detect the intension from the matrix dimensions :thumbsup:

**Reply by dvd101x** [reply]
### 1. broadcast_to(A, dimensions)

I agree it might be better for the user to include something like repeat. bradcast_to is more like a building block for broadcasting but I think it would not be used elsewhere. Something like repeat on the other hand could be more useful in general and could be referenced from `concat`

Here are some references for something like repeat.
- [Numpy: repeat](https://numpy.org/doc/stable/reference/generated/numpy.repeat.html)
- [Octave: repmat](https://octave.sourceforge.io/octave/function/repmat.html)

### 2. broadcast_matrices(A, B)

Yes I agree it should be best to start simple, maybe it could also be a building block for broadcasting, but not available for the user. I imagine these algorithms in order of more simple to more efficient.

1. As is, repeat the matrices until all matrices are the same size (only if the dimension to repeat is none or one) and then use the existing algorithm that iterates in matrices element wise.
2. Reshape the matrices and modify the algorithm that iterates in matrices to find dimensions 1 if dimension doesn't match N. Thus a new matrix is created but not repeated N times.
3. Keep the matrices as they are and modify the algorithm that iterates in matrices to find not existing dimensions, dimensions 1 if dimension doesn't match N. Thus the same matrices are used and not copied.

I think it's best to start with the basics like `add`, `subtract`, `dotMultiply`, `dotDivide` and `dotPow` to accommodate some of the interest for Machine Learning in the browser. 

If it all works, maybe in a another PR include other functions that seem straightforwards in groups. Like `and`, `not`, `or`, `xor` as I think it would only be a matter of broadcasting all inputs.

And lastly check some others that takes different kinds of inputs and depend on the order.

### 3. add(a, b, ~dimension~)

👍

**Reply by josdejong** [reply]
Makes sense to start this broadcast function as an internal function, and only expose the end result via a function like `repeat` and more flexible operators/functions like `add` and `substract`. 

It sounds good to start with `add`, `subtract`, `dotMultiply`, `dotDivide` and `dotPow`.

Is this something you plan to implement and contribute to mathjs? (that would be awesome)

**Reply by dvd101x** [reply]
Yes, I would love to contribute with it!

I might take a while since I don't know how to work with modules correctly. Even though I have read [build](https://github.com/josdejong/mathjs#build) and [develop](https://github.com/josdejong/mathjs#develop) I'm not familiar with npm. I can take this as a learning opportunity.

I think I can test in the browser at first and iterate until it's acceptable.

**Reply by josdejong** [reply]
Thanks! Just let me know if you need help.

**Reply by dvd101x** [reply]
Hi, @josdejong 

Here is an update and a question

I was able to add my function `broadcast`, build and validate it works.

 So now it works as `math.broadcast`

I was able to test most candidates for broadcasting using imports with broadcasting and they work ok on this project.
https://dvd101x.github.io/web-thermodynamics/

Now I would like to tweak `add` on source so it uses broadcast first. It seems that I will either need to:
- modify [matrixAlgorithmSuite](https://github.com/josdejong/mathjs/blob/develop/src/type/matrix/utils/matrixAlgorithmSuite.js) which would affect everything
- convert the current `add` to `addWithoutBroadcasting` and then call it from a function `add` that uses `addWithoutBroadcasting(broadcast(...matrices))` (this would have a redundancy with the current broadcasting with scalars)

So I would like some advice. ¿What would be an acceptable way of modifying add so they use first the `broadcast` function?

> **PD1:** regarding the function `repeat` I found a good reference on another project but it's broader than what I can apply right now. So I used a very specific function called `stretch` that is only used inside broadcasting and could later be replaced by `repeat`
> **PD2:** In the end I would like to leave broadcast as an util that can be used by element wise functions but I'm testing it as a function until I can call it correctly

**Reply by dvd101x** [reply]
Update, I made `broadcastedMatrixAlgorithmSuite` that is like [matrixAlgorithmSuite](https://github.com/josdejong/mathjs/blob/develop/src/type/matrix/utils/matrixAlgorithmSuite.js) but broadcasts inputs first. It seems to be less intrusive than the alternatives I was implying.

**Reply by dvd101x** [reply]
Hi, I'm sorry about the frequent postings, it's just an update on how things are going.

I did some manual tests and it seems to work for `add`, `subtract`, `dotMultiply`, `dotDivide` and `dotPow`.

I think there is something wrong in my broadcast function due to some helper functions I use that messes up system testing.

[src/type/matrix/utils/broadcast.js](https://github.com/dvd101x/mathjs/blob/9232a9801785a5ce66f649801ecd5c4cd3513328/src/type/matrix/utils/broadcast.js)

Could you give me some pointers on how to improve this?

I don't know if I should export my helper functions to factoriesAny (even though I don't think they will be useful outside broadcasting)

**Reply by dvd101x** [reply]
Hi, I found a way to simplify `broadcast` to be used specifically by `matrixAlgorithmSuite` and will report back at the end of the week, it seems promising.

**Reply by josdejong** [reply]
That sounds promising. It makes sense to extend the `matrixAlgorithmSuite` for this, and the nice thing is that we will then implement broadcasting for all relevant functions at once :) . If we want to be careful with that, we could add a flag to the `matrixAlgorithmSuite` function whether or not we want broadcasting. Then we can turn it on for functions one by one and add unit tests. Not sure though.

So you're implementing the broadcast function as a new public function, not an internal function? I think that is fine, though it requires more documentation, testing and thinking through of the API. The easy way would be to keep it an internal function for now. Both approaches are fine with me.

I understand from your last message that you're on track and earlier questions are resolved.

**Reply by dvd101x** [reply]
Hi Jos,

Yes my early questions are now resolved.

I ended with an internal function `broadcast` that seems very promising. It's already working in my fork, but tests are failing since now the errors for missmatched sizes are working as expected, I will modify the tests so they fail according to the rules of broadcasting.

Now I see a few options:

- Change all at once by modifying `matrixAlgorithmSuite`
- Change a few by using `broadcastedMatrixAlgorithmSuite` and after the last function is done, eliminate `broadcastedMatrixAlgorithmSuite` and only use a broadcasted version of `matrixAlgorithmSuite`
- Add a flag to `matrixAlgorithmSuite` but I don't know how

**Reply by dvd101x** [reply]
I modified a few tests and it seems to work now for the basic arithmetic functions for `add`, `subtract`, `dotMultiply`, `dotDivide` and `dotPow`.

**Reply by dvd101x** [reply]
Hi Jos,

I did implement broadcast with an internal function in `matrixAlgorithmSuite` and it works nicely for the full suite of functions.

**Reply by josdejong** [reply]
That sounds good! Nicest would be changing all at once. We should properly unit test the new functionality though and have full confidence that it works OK everywhere. Without seeing the actual code I'm not entirely sure, but I think that will work out just fine :)

**Reply by dvd101x** [reply]
Thanks. I have it in my repository is it ok if I make a pull request?

**Reply by josdejong** [reply]
Yes please!

**Reply by dvd101x** [reply]
I included pull request #2895 regarding this topic.

**Reply by josdejong** [reply]
Published now in `v11.6.0` 🎉

**Reply by dvd101x** [reply]
Thanks!

**Comment by ChristopherChudzicki** [comment]
This could definitely be useful in some cases. But  I suggest that there should be a way to turn it off.

Example: I mainly use MathJS for math webapps to be used by high-school and undergraduate students. I generally want students to think that adding matrix + vector is not valid. Great, MathJS throws a useful error!

I could be one of a small minority, but I would guess that MathJS's current behavior is useful to other people.

(Admittedly, if people like me want this error, it's not too hard to do our own validation. But that makes upgrading harder.)

Incidentally, `numpy` calls this  [broadcasting](https://docs.scipy.org/doc/numpy/user/basics.broadcasting.htm). (Although, Numpy's broadcasting is more general.)

Edit: To be clear, doing my own validation really is not a big deal. Just something to consider :)

**Comment by JasonShin** [comment]
Thanks for the inputs guys =)

**Comment by josdejong** [comment]
Thanks for sharing your thoughts @ChristopherChudzicki. Very good point to weight convenience against more explicit notations preventing you from accidental mistakes. We could also think about a solution where we have separate, explicit functions (like `subtractVector(matrix, vector)`) to do operations on matrices and vectors, or pass an extra option to these functions 
(like `subtract(matrix, vector, { broadcast: true })` 
or `subtract(matrix, vector, { rowWise: true })`). 

Just thinking aloud here :)

**Comment by ChristopherChudzicki** [comment]
That said, it occurs to me that allowing `matrix + row` or `matrix + vector` consistent with how MathJS handles `matrix + scalar`. What I mean is, in the `matrix+scalar` case MathJS already does not demand strict shape-matching.

So maybe this is the "right" behavior.

**Comment by josdejong** [comment]
Yes that's true.

I'm just exploring different options, not sure yet what solution would best serve the audience of mathjs.

**Comment by flowstatelabs** [comment]
Any progress on this? Am getting into mathjs as a numpy replacement in the browser for implementing ML/AI stuffs. [Broadcasting](https://docs.scipy.org/doc/numpy/user/basics.broadcasting.html) is a verrry handy feature that makes a lot of learning algorithms a bit simpler to implement. Happy to pitch in.

**Comment by flowstatelabs** [comment]
Also might add the word "Broadcasting" to the title for the numpy folks looking for help on this. I'd imagine they'll become a larger and larger part of your target audience, if they aren't currently.

**Comment by josdejong** [comment]
@chivington thanks for your input. There is no progress on this feature so far. I've marked the issue with a label "help wanted", now.

Anyone interested in implementing this feature?

**Comment by joeGermany** [comment]
Hello!! My case is similar to @chivington. I am getting into machine learning on the web and I need an alternative for NumPy. The above feature (broadcasting) is a real help for easier manipulation of matrices. Is there any way I can implement it alone in the meantime until it is a mainstream MathJS function (if it will ever be). 😃

**Comment by josdejong** [comment]
> Is there any way I can implement it alone in the meantime until it is a mainstream MathJS function (if it will ever be). 

The solution would be to implement your own version of all operators/functions like subtract, add, etc, like Jason suggests in https://github.com/josdejong/mathjs/issues/1165#issuecomment-404493249.

**Comment by gwhitney** [comment]
The basic functionality proposed here was implemented as of #2895, so closing this discussion.
